third_party/dml/ci/generate_absl_test_summary.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""
+Parses an AbslTest XML log to produce a flattened summary of test results with the following structure:
+{
+    "Counts":
+    {
+        "Passed": [int]
+        "Failed": [int]
+        "Skipped": [int]
+        "Blocked": [int]
+        "Total": [int]
+    }
+
+    "Tests":
+    [
+        {
+            "Name": [string]
+            "Result": [string] (one of "Passed", "Failed", "Skipped", "Blocked")
+            "Time" : [double] (time to execute in seconds)
+            "Errors": [string] (optional; only present if result is "fail")
+        }
+        ...
+    ]
+
+    "Time" : [double] (time for all modules to finish, in seconds; can be smaller than the sum of test times, because of parallel execution)
+    "Errors": [string] (optional; only present if errors occur outside of test scopes)
+}
+"""
+
+import collections
+import re
+import argparse
+import xml.etree.ElementTree as ET
+import json
+import glob
+import os
+
+CrashedTestInfo = collections.namedtuple('CrashedTestInfo', 'name path error')
+
+
+def _parse_args():
+  """Parses the arguments given to this script."""
+
+  parser = argparse.ArgumentParser()
+  parser.add_argument("--log_path",
+                      help="path of the test's text file log",
+                      required=True)
+
+  parser.add_argument("--out_summary_path",
+                      help="path to the summary json file to output",
+                      required=True)
+
+  parser.add_argument("--out_error_log_path",
+                      help="path to the error log to output",
+                      required=True)
+
+  parser.add_argument("--xml_files_dir",
+                      help="paths of directory where the xml files are located",
+                      required=True)
+
+  parser.add_argument("--start_time",
+                      help="the start time of the tests",
+                      required=True)
+
+  parser.add_argument("--end_time",
+                      help="the end time of the tests",
+                      required=True)
+
+  return parser.parse_args()
+
+
+def _parse_test_crashes(log_path):
+  test_crashes = []
+
+  path_pattern = re.compile(r"Running '(.+)' with shard \d+\.\.\.")
+
+  with open(log_path, 'r', encoding='utf-16') as test_log:
+    line = test_log.readline()
+
+    current_test_title = None
+    current_test_path = ''
+
+    while line:
+      if line.startswith('[ RUN      ]'):
```

Comment:
How certain are we that "[ RUN      ]" will always be 12 characters? Should it be a pattern instead, like "[ *RUN *]"?

Reply:
It's always been 12 characters (if it ever changes then it most likely won't be named "RUN" anymore)

Reply:
gtest v1.12.0 "[ SEE SPOT RUN ]" 😅